sdks/v4-sdk/src/entities/pool.ts

```diff
@@ -0,0 +1,259 @@
+import { ADDRESS_ZERO, NEGATIVE_ONE, Q192 } from '../utils/internalConstants'
+import JSBI from 'jsbi'
+
+const NO_TICK_DATA_PROVIDER_DEFAULT = new NoTickDataProvider()
+
+/**
+ * Represents a V4 pool
+ */
+export class Pool {
+  public readonly currency0: Currency
+  public readonly currency1: Currency
+  public readonly fee: number
+  public readonly tickSpacing: number
+  public readonly sqrtRatioX96: JSBI
+  public readonly hooks: string // address
+  public readonly liquidity: JSBI
+  public readonly tickCurrent: number
+  public readonly tickDataProvider: TickDataProvider
+
+  private _currency0Price?: Price<Currency, Currency>
+  private _currency1Price?: Price<Currency, Currency>
+
+  public static getPoolId(
+    currencyA: Currency,
+    currencyB: Currency,
+    fee: number,
+    tickSpacing: number,
+    hooks: string
+  ): string {
+    const [currency0, currency1] = sortsBefore(currencyA, currencyB) ? [currencyA, currencyB] : [currencyB, currencyA]
+    const currency0Addr = currency0.isNative ? ADDRESS_ZERO : currency0.wrapped.address
+    const currency1Addr = currency1.isNative ? ADDRESS_ZERO : currency1.wrapped.address
+    return keccak256(
+      ['bytes'],
+      [
+        defaultAbiCoder.encode(
+          ['address', 'address', 'uint24', 'int24', 'address'],
+          [currency0Addr, currency1Addr, fee, tickSpacing, hooks]
+        ),
+      ]
+    )
+  }
+
+  /**
+   * Construct a pool
+   * @param currencyA One of the currencys in the pool
+   * @param currencyB The other currency in the pool
+   * @param fee The fee in hundredths of a bips of the input amount of every swap that is collected by the pool
+   * @param sqrtRatioX96 The sqrt of the current ratio of amounts of currency1 to currency0
+   * @param liquidity The current value of in range liquidity
+   * @param tickCurrent The current tick of the pool
+   */
+  public constructor(
+    currencyA: Currency,
+    currencyB: Currency,
+    fee: number,
+    tickSpacing: number,
+    hooks: string,
+    sqrtRatioX96: BigintIsh,
+    liquidity: BigintIsh,
+    tickCurrent: number,
+    ticks: TickDataProvider | (Tick | TickConstructorArgs)[] = NO_TICK_DATA_PROVIDER_DEFAULT
+  ) {
+    invariant(Number.isInteger(fee) && fee < 1_000_000, 'FEE')
+    const tickCurrentSqrtRatioX96 = TickMath.getSqrtRatioAtTick(tickCurrent)
+    const nextTickSqrtRatioX96 = TickMath.getSqrtRatioAtTick(tickCurrent + 1)
+    invariant(
+      JSBI.greaterThanOrEqual(JSBI.BigInt(sqrtRatioX96), tickCurrentSqrtRatioX96) &&
+        JSBI.lessThanOrEqual(JSBI.BigInt(sqrtRatioX96), nextTickSqrtRatioX96),
+      'PRICE_BOUNDS'
+    )
+    // always create a copy of the list since we want the pool's tick list to be immutable
+    ;[this.currency0, this.currency1] = sortsBefore(currencyA, currencyB)
+      ? [currencyA, currencyB]
+      : [currencyB, currencyA]
+    this.fee = fee
+    this.sqrtRatioX96 = JSBI.BigInt(sqrtRatioX96)
+    this.tickSpacing = tickSpacing
+    this.hooks = hooks
```

Comment:
nit: we probably want to validate hooks address during constructor, [checkValidAddress](https://github.com/Uniswap/sdks/blob/main/sdks/sdk-core/src/utils/validateAndParseAddress.ts#L22C17-L22C34).  I'm unsure about ethers library behavior in case of zero address though.